File: python/scio/io.py
"""
Main I/O module - Simple API for writing and reading AnnData in tar format
"""
import anndata as ad
import tarfile
import tempfile
import shutil
from pathlib import Path
from typing import Union

from .writers import save_to_folder, update_folder
from .readers import load_from_folder
from .formats import load_json
from .lazy import LazyAnnData


def write(adata: ad.AnnData,
          path: Union[str, Path],
          sparse_format: str = 'csr',
          overwrite: bool = False,
          update: bool = False) -> None:
    """
    Write AnnData object to .scio folder

    Saves ALL 10 components of AnnData:
    - X (expression matrix)
    - obs (cell metadata)
    - var (gene metadata)
    - obsm (cell embeddings: PCA, UMAP, etc.)
    - varm (gene embeddings)
    - obsp (cell-cell graphs)
    - varp (gene-gene graphs)
    - layers (additional matrices)
    - raw (raw counts)
    - uns (unstructured metadata)

    Format inside .scio (v0.1.3):
    - Binary CSC/CSR (.npy) for sparse matrices (fast, cross-platform)
    - Parquet for DataFrames (preserves dtypes, efficient for large data)
    - JSON for metadata

    Parameters:
    -----------
    adata : AnnData
        AnnData object to save
    path : str or Path
        Output .scio folder path (e.g., "data.scio")
    sparse_format : str
        Format for sparse matrices: 'csr' (default, fastest write from scanpy)
        or 'csc' (faster R read). Default is 'csr' since scanpy outputs CSR.
    overwrite : bool
        Whether to overwrite existing folder (default: False)
    update : bool
        Whether to perform incremental update using hash-based change detection
        (default: False). Only changed components will be rewritten.

    Example:
    --------
    >>> import anndata as ad
    >>> import scio
    >>> adata = ad.read_h5ad("data.h5ad")
    >>> scio.write(adata, "data.scio")
    >>>
    >>> # Use CSC format for faster R reading
    >>> scio.write(adata, "data.scio", sparse_format='csc')
    >>>
    >>> # Incremental update (only changed components)
    >>> scio.write(adata, "data.scio", update=True)
    """
    path = Path(path)

    # Validate sparse_format
    if sparse_format not in ('csr', 'csc'):
        raise ValueError(f"sparse_format must be 'csr' or 'csc', got '{sparse_format}'")

    # Validate extension
    if not str(path).endswith('.scio'):
        raise ValueError(
            f"Invalid file extension. Only '.scio' is supported.\n"
            f"Got: {path.suffix}\n"
            f"Example: scio.write(adata, 'data.scio')"
        )

    # Validate conflicting options
    if overwrite and update:
        raise ValueError("Cannot use both overwrite=True and update=True. Choose one.")

    # Handle update mode
    if update and path.exists():
        manifest_path = path / "manifest.json"
        if not manifest_path.exists():
            raise ValueError(f"Invalid scio folder: manifest.json not found in {path}")

        manifest = load_json(manifest_path)

        if not manifest.get('hashes'):
            # No hash info - do full save with hashes
            print("  No hash info found in existing file. Doing full save with hash computation...", flush=True)
            save_to_folder(adata, path, sparse_format=sparse_format, compute_hashes=True)
            print(f"  ✓ Data saved with hashes to {path.name}", flush=True)
        else:
            # Incremental update - only write changed components
            print("  Performing incremental update...", flush=True)
            update_folder(adata, path, manifest, sparse_format=sparse_format)
            print(f"  ✓ Incremental update complete for {path.name}", flush=True)

        return

    # Check if folder exists for non-update mode
    if path.exists() and not overwrite and not update:
        raise FileExistsError(
            f"Folder already exists: {path}\n"
            f"Use overwrite=True to replace it, or update=True for incremental update."
        )

    # Remove existing folder if overwrite is True
    if path.exists() and overwrite:
        shutil.rmtree(path)

    print(f"  Saving data to {path.name}...", flush=True)
    # Compute hashes if update=True (for future incremental updates)
    # Skip hash computation on normal writes for speed
    save_to_folder(adata, path, sparse_format=sparse_format, compute_hashes=update)
    print(f"  ✓ Data saved to {path.name}", flush=True)


def read(path: Union[str, Path], lazy: bool = False) -> Union[ad.AnnData, LazyAnnData]:
    """
    Read AnnData object from .scio folder (or legacy tar archive)

    Loads ALL 10 components of AnnData:
    - X (expression matrix)
    - obs (cell metadata)
    - var (gene metadata)
    - obsm (cell embeddings)
    - varm (gene embeddings)
    - obsp (cell-cell graphs)
    - varp (gene-gene graphs)
    - layers (additional matrices)
    - raw (raw counts)
    - uns (unstructured metadata)

    Parameters:
    -----------
    path : str or Path
        Path to .scio folder (also supports legacy tar archives)
    lazy : bool
        If True, use lazy loading (only load components when accessed).
        This significantly improves performance for large datasets.
        Default: False (load all data immediately)

    Returns:
    --------
    AnnData or LazyAnnData :
        Reconstructed AnnData object (lazy=False) or
        LazyAnnData object (lazy=True)

    Example:
    --------
    >>> import scio
    >>> # Standard loading - loads all data
    >>> adata = scio.read("data.scio")
    >>>
    >>> # Lazy loading - only loads structure initially
    >>> adata = scio.read("data.scio", lazy=True)
    >>> print(adata.shape)  # Fast - no data loaded
    >>> cell_types = adata.obs['cell_type']  # Loads obs on first access
    >>> expression = adata.X  # Loads X on first access
    >>>
    >>> # Convert lazy to full AnnData when needed
    >>> full_adata = adata.to_memory()
    """
    path = Path(path)

    if not path.exists():
        raise FileNotFoundError(f"File not found: {path}")

    # Check if it's a folder or a tar archive
    if path.is_dir():
        # It's a folder - load directly (or lazy load)
        if lazy:
            print(f"  Initializing lazy loading from {path.name}...", flush=True)
            adata = LazyAnnData(path)
            print(f"  ✓ Structure loaded (data will load on access)", flush=True)
            return adata
        else:
            print(f"  Loading data from {path.name}...", flush=True)
            adata = load_from_folder(path)
            print(f"  ✓ Data loaded", flush=True)
            return adata

    # It's a file - try to open as tar archive (for backward compatibility)
    if lazy:
        raise ValueError(
            "Lazy loading is not supported for legacy tar archives.\n"
            "Please use the folder format (.scio) for lazy loading."
        )

    try:
        print(f"  Extracting legacy tar archive...", flush=True)
        with tempfile.TemporaryDirectory() as tmpdir:
            tmpdir = Path(tmpdir)

            # Extract tar archive
            with tarfile.open(path, 'r') as tar:
                tar.extractall(tmpdir)

            # Find the extracted folder (should be only one)
            extracted_folders = [d for d in tmpdir.iterdir() if d.is_dir()]

            if len(extracted_folders) == 0:
                raise ValueError(f"No folder found in archive: {path}")
            elif len(extracted_folders) > 1:
                raise ValueError(f"Multiple folders found in archive: {path}")

            data_folder = extracted_folders[0]

            # Load from folder structure
            print(f"  Loading data...", flush=True)
            adata = load_from_folder(data_folder)
            print(f"  ✓ Data loaded", flush=True)

        return adata

    except tarfile.ReadError as e:
        raise ValueError(
            f"Path is not a valid .scio folder or tar archive: {path}\n"
            f"Expected either a .scio folder or a legacy tar archive.\n"
            f"Error: {str(e)}"
        )

File: R/R/io.R
# Main I/O module - Simple API for writing and reading data in .scio format
# Part of the scio R package

#' Write single-cell data to .scio folder
#'
#' Saves Seurat or SingleCellExperiment objects to .scio folder format
#' using binary sparse format for sparse matrices.
#' Works seamlessly with Python (loads as AnnData).
#'
#' v0.1.3: Supports both CSR and CSC sparse matrix formats (.npy files).
#' Data is stored in cells x genes orientation.
#'
#' @param object Seurat or SingleCellExperiment object
#' @param path Output .scio folder path (e.g., "data.scio")
#' @param sparse_format Format for sparse matrices: "csr" (default, fastest write)
#'   or "csc" (faster R read). Default is "csr" since scanpy outputs CSR.
#' @param overwrite Whether to overwrite existing folder (default: FALSE)
#' @param update Whether to perform incremental update using hash-based
#'   change detection (default: FALSE). Only changed components will be
#'   rewritten. If the existing file has no hash info, a full save is
#'   performed with hashes computed for future updates.
#'
#' @examples
#' \dontrun{
#' library(scio)
#'
#' # Save Seurat object (full write, default CSR format)
#' scio_write(seurat_obj, "data.scio")
#'
#' # Use CSC format for faster R reading
#' scio_write(seurat_obj, "data.scio", sparse_format = "csc")
#'
#' # Overwrite existing file
#' scio_write(seurat_obj, "data.scio", overwrite = TRUE)
#'
#' # Incremental update (only changed components)
#' scio_write(seurat_obj, "data.scio", update = TRUE)
#' }
#'
#' @export
scio_write <- function(object, path, sparse_format = "csr", overwrite = FALSE, update = FALSE) {
  # Validate sparse_format
  sparse_format <- match.arg(sparse_format, c("csr", "csc"))

  # Validate conflicting options
  if (overwrite && update) {
    stop("Cannot use both overwrite = TRUE and update = TRUE. Choose one.")
  }

  # Detect object type and extract components
  if (inherits(object, "Seurat")) {
    components <- extract_components_from_seurat(object)
  } else if (inherits(object, "SingleCellExperiment")) {
    components <- extract_components_from_sce(object)
  } else {
    stop("Unsupported object type. Must be Seurat or SingleCellExperiment.")
  }

  # Handle update mode
  if (update && dir.exists(path)) {
    # Load existing manifest
    manifest_path <- file.path(path, "manifest.json")
    if (!file.exists(manifest_path)) {
      stop(paste0("Invalid scio folder: manifest.json not found in ", path))
    }

    manifest <- load_json(manifest_path)

    if (is.null(manifest$hashes)) {
      # No hash info - do full save with hashes
      message("  No hash info found in existing file. Doing full save with hash computation...")
      save_to_folder(components, path, sparse_format = sparse_format, compute_hashes = TRUE)
      message("  \u2713 Data saved with hashes to ", basename(path))
    } else {
      # Incremental update - only write changed components
      message("  Performing incremental update...")
      update_folder(components, path, manifest, sparse_format = sparse_format)
      message("  \u2713 Incremental update complete for ", basename(path))
    }

    return(invisible(path))
  }

  # Check if folder exists for non-update mode
  if (dir.exists(path) && !overwrite && !update) {
    stop(paste0("Folder already exists: ", path, "\nUse overwrite = TRUE to replace it, or update = TRUE for incremental update."))
  }

  # Remove existing folder if overwrite is TRUE
  if (dir.exists(path) && overwrite) {
    unlink(path, recursive = TRUE)
  }

  # Save directly to .scio folder (skip hash computation for faster write)
  message("  Saving data to ", basename(path), "...")
  save_to_folder(components, path, sparse_format = sparse_format, compute_hashes = FALSE)
  message("  \u2713 Data saved to ", basename(path))

  invisible(path)
}


#' Read single-cell data from .scio folder
#'
#' Loads data from .scio folder and converts to Seurat or SingleCellExperiment.
#'
#' @param path Path to .scio folder (also supports legacy tar archives)
#' @param output Output format: "Seurat" or "SCE" (default: "Seurat")
#'
#' @return Seurat or SingleCellExperiment object
#'
#' @examples
#' \dontrun{
#' library(scio)
#'
#' # Load as Seurat
#' seurat_obj <- scio_read("data.scio", output = "Seurat")
#'
#' # Load as SingleCellExperiment
#' sce_obj <- scio_read("data.scio", output = "SCE")
#' }
#'
#' @export
scio_read <- function(path, output = "Seurat") {
  if (!file.exists(path)) {
    stop(paste0("File not found: ", path))
  }

  # Validate output format
  output <- match.arg(output, c("Seurat", "SCE"))

  # Check if it's a folder or a tar archive
  if (dir.exists(path)) {
    # It's a folder - load directly
    message("  Loading data from ", basename(path), "...")
    components <- load_from_folder(path)
    message("  \u2713 Data loaded")
  } else {
    # It's a file - try tar archive (for backward compatibility)
    message("  Extracting legacy tar archive...")
    tmpdir <- tempdir()

    # Extract tar archive
    untar(path, exdir = tmpdir)

    # Find the extracted folder
    folder_name <- tools::file_path_sans_ext(basename(path))
    data_folder <- file.path(tmpdir, folder_name)

    if (!dir.exists(data_folder)) {
      stop(paste0("Extracted folder not found: ", data_folder))
    }

    # Load from folder structure
    message("  Loading data...")
    components <- load_from_folder(data_folder)
    message("  \u2713 Data loaded")

    # Clean up temporary folder
    unlink(data_folder, recursive = TRUE)
  }

  # Convert to requested format
  if (output == "Seurat") {
    result <- create_seurat_from_components(components)
  } else if (output == "SCE") {
    result <- create_sce_from_components(components)
  }

  return(result)
}


#' Write components list to .scio folder (advanced usage)
#'
#' @param components Named list with all components
#' @param path Output .scio folder path
#' @param overwrite Whether to overwrite existing folder
#' @export
scio_write_components <- function(components, path, overwrite = FALSE) {
  # Check if folder exists
  if (dir.exists(path) && !overwrite) {
    stop(paste0("Folder already exists: ", path, "\nUse overwrite = TRUE to replace it."))
  }

  # Remove if exists
  if (dir.exists(path) && overwrite) {
    unlink(path, recursive = TRUE)
  }

  # Save to folder structure (skip hash computation for faster write)
  save_to_folder(components, path, compute_hashes = FALSE)

  invisible(path)
}


#' Read components list from .scio folder (advanced usage)
#'
#' @param path Path to .scio folder
#' @return Named list with all components
#' @export
scio_read_components <- function(path) {
  if (!file.exists(path)) {
    stop(paste0("Path not found: ", path))
  }

  if (dir.exists(path)) {
    # It's a folder - load directly
    components <- load_from_folder(path)
  } else {
    # Legacy tar archive support
    tmpdir <- tempdir()
    untar(path, exdir = tmpdir)
    folder_name <- tools::file_path_sans_ext(basename(path))
    data_folder <- file.path(tmpdir, folder_name)

    if (!dir.exists(data_folder)) {
      stop(paste0("Extracted folder not found: ", data_folder))
    }

    components <- load_from_folder(data_folder)
    unlink(data_folder, recursive = TRUE)
  }

  return(components)
}
